### PR TITLE
Composite electrode model

### PR DESCRIPTION
**Pull request by weilongai**
# Description

Add a composite electrode particle model for composite electrodes with two phases. e.g. graphite/silicon for the LG M50. This model includes the basic functions, and users are encouraged to add more features by adapting it to sub-models.

This work is referred to a recent paper below
W. Ai, N. Kirkaldy, Y. Jiang, G. Offer, H. Wang, B. Wu (2022). A composite electrode model for lithium-ion battery with a silicon/graphite negative electrode. Journal of Power Sources. 527, 231142.

Fixes # (1433)

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [ ] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

## Actions involved
- A new notebook  [How to use composite particle model] (https://github.com/weilongai/PyBaMM/blob/composite-release/examples/notebooks/models/composite_particle.ipynb)
- A new basic function [basic_dfn_composite.py](https://github.com/weilongai/PyBaMM/blob/composite-release/pybamm/models/full_battery_models/lithium_ion/basic_dfn_composite.py)
- A new set of parameters for the composite electrode materials [graphite_Chen2020_composite](https://github.com/weilongai/PyBaMM/tree/composite-release/pybamm/input/parameters/lithium_ion/negative_electrodes/graphite_Chen2020_composite)
- A new test function [test_dfn_composite.py](https://github.com/weilongai/PyBaMM/blob/composite-release/tests/unit/test_models/test_full_battery_models/test_lithium_ion/test_dfn_composite.py)
- Modifications to the [lithium_ion_parameters.py](https://github.com/weilongai/PyBaMM/blob/composite-release/pybamm/parameters/lithium_ion_parameters.py)

### CODE REVIEW

**Comment by weilongai** [comment]
There is an error in the test function test_basic_models.py, but it is unclear what are the reasons.

**Comment by rtimms** [comment]
It's because in `basic_dfn.py` the variables name still has the word "of" in it. See [here](https://github.com/pybamm-team/PyBaMM/pull/1949/files#diff-fb7fbe7f25ca70b2a4d3f5edd5198eaca742e5c6d20a7ac3d40297cff601ec15R301).

**Comment by weilongai** [comment]
> It's because in `basic_dfn.py` the variables name still has the word "of" in it. See [here](https://github.com/pybamm-team/PyBaMM/pull/1949/files#diff-fb7fbe7f25ca70b2a4d3f5edd5198eaca742e5c6d20a7ac3d40297cff601ec15R301).

Thanks, Robert. Another issue appears (see below), but it seems not to be from this branch.
```
FAIL: test_compare_dfns (test_models.test_full_battery_models.test_lithium_ion.test_compare_basic_models.TestCompareBasicModels)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/PyBaMM/PyBaMM/tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_compare_basic_models.py", line 31, in test_compare_dfns
    np.testing.assert_array_almost_equal(
  File "/home/runner/work/PyBaMM/PyBaMM/.tox/integration/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 1046, in assert_array_almost_equal
    assert_array_compare(compare, x, y, err_msg=err_msg, verbose=verbose,
  File "/home/runner/work/PyBaMM/PyBaMM/.tox/integration/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 844, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not almost equal to 4 decimals
```

**Comment by tinosulzer** [comment]
Are you planning another PR where the composite electrode model is added as a submodel?

**Comment by weilongai** [comment]
> Are you planning another PR where the composite electrode model is added as a submodel?

No, I don't have a plan for submodels now. The basic functions serve to explain the interactions between silicon and graphite. If someone would like to study more complex chemistry (I guess there are plans already), he (or she) is encouraged to convert the basic functions to submodels and then to add more features.

File: examples/notebooks/Getting Started/Tutorial 1 - How to run a model.ipynb [hunk 1]
```diff
@@ -278,7 +278,11 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
+<<<<<<< HEAD
```
Comment (tinosulzer) [inline examples/notebooks/Getting Started/Tutorial 1 - How to run a model.ipynb:281]: fix conflicts in notebooks

File: examples/notebooks/parameterization/change-input-current.ipynb [hunk 2]
```diff
@@ -351,7 +351,11 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
+<<<<<<< HEAD:examples/notebooks/change-input-current.ipynb
```
Comment (tinosulzer) [inline examples/notebooks/parameterization/change-input-current.ipynb:354]: fix conflicts

File: pybamm/expression_tree/operations/evaluate_python.py [hunk 3]
```diff
@@ -69,12 +69,15 @@ def dot_product(self, b):
...
     def scalar_multiply(self, b):
         """
-        multiply of matrix with a scalar b
+                multiply of matrix with a scalar b
 
-        Parameters
-        ----------
-        b: Number or 1 element jax device array
-            scalar value to multiply
+        <<<<<<< HEAD:pybamm/expression_tree/operations/evaluate.py
```
Comment (tinosulzer) [inline pybamm/expression_tree/operations/evaluate_python.py:74]: fix conflicts

File: pybamm/expression_tree/operations/evaluate_python.py [hunk 4]
```diff
@@ -92,13 +95,14 @@ def __matmul__(self, b):
...
 def create_jax_coo_matrix(value):
     """
-    Creates a JaxCooMatrix from a scipy.sparse matrix
+        Creates a JaxCooMatrix from a scipy.sparse matrix
 
-    Parameters
-    ----------
+        Parameters
+        ----------
+    >>>>>>> develop:pybamm/expression_tree/operations/evaluate_python.py
```
Comment (tinosulzer) [inline pybamm/expression_tree/operations/evaluate_python.py:102]: fix conflicts

File: pybamm/models/submodels/external_circuit/current_control_external_circuit.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
+#
```
Comment (tinosulzer) [inline pybamm/models/submodels/external_circuit/current_control_external_circuit.py:1]: you can remove this file

File: pybamm/models/submodels/interface/base_interface.py [hunk 6]
```diff
@@ -141,15 +141,17 @@ def _get_exchange_current_density(self, variables):
...
+    def _get_open_circuit_potential(self, variables, phase=None):
         """
         A private function to obtain the open circuit potential and entropic change
 
         Parameters
         ----------
         variables: dict
             The variables in the full model.
-
+        phase : string
```
Comment (tinosulzer) [inline pybamm/models/submodels/interface/base_interface.py:152]: Using `1` or `2` doesn't give much information. Could you use more informative names ('graphite' and 'silicon')?

File: pybamm/parameters/lithium_ion_parameters.py [hunk 7]
```diff
@@ -394,6 +394,26 @@ def _set_dimensional_parameters(self):
             "Positive electrode reaction-driven LAM factor [m3.mol-1]"
         )
 
+        # composite particle
+        self.c_n_p1_max = pybamm.Parameter(
+            "Maximum concentration in negative electrode of phase 1 [mol.m-3]"
```
Comment (tinosulzer) [inline pybamm/parameters/lithium_ion_parameters.py:399]: use more informative names

File: tests/unit/test_models/test_submodels/standard_submodel_unit_tests.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
+#
```
Comment (tinosulzer) [inline tests/unit/test_models/test_submodels/standard_submodel_unit_tests.py:1]: why was this file added?

**Review by tinosulzer — CHANGES_REQUESTED** [review]
I'm disappointed that you've only added a basic model and not a submodel for this. It goes against the spirit of the common modeling framework, which is to be able to couple different submodels. Now this model can't be used with thermal effects, experiments, degradation, etc. I know it might not be obvious how you would add this as a submodel, but we could have helped you, and you're now asking someone to do this entirely for you with no incentive, when you could have asked earlier and someone would have helped in return for co-authorship on your paper.

File: pybamm/models/submodels/interface/base_interface.py (hunk 6 above)
Comment (weilongai) [reply pybamm/models/submodels/interface/base_interface.py:152]: They are general names, not only for graphite and silicon

**Comment by weilongai** [comment]
> I'm disappointed that you've only added a basic model and not a submodel for this. It goes against the spirit of the common modeling framework, which is to be able to couple different submodels. Now this model can't be used with thermal effects, experiments, degradation, etc. I know it might not be obvious how you would add this as a submodel, but we could have helped you, and you're now asking someone to do this entirely for you with no incentive, when you could have asked earlier and someone would have helped in return for co-authorship on your paper.

I did raise an issue  #1433, but found that it would take much more time to make it submodel. There are similar papers appearing (see https://doi.org/10.1149/1945-7111/ac5481 and http://dx.doi.org/10.1149/1945-7111/ac51f3), so it was in a rush. This follows a solution-led strategy, i.e. using the simplest way (basic functions) to solve the problem. When people are studying new problems (adding more interactions), it is straight to adapt this model to meet their requirements, which are unknown at this stage.

File: tests/unit/test_models/test_submodels/standard_submodel_unit_tests.py (hunk 8 above)
Comment (weilongai) [reply tests/unit/test_models/test_submodels/standard_submodel_unit_tests.py:1]: I did not add this. It may be from an old branch. If you agree, I can remove it or leave it to you.

**Comment by weilongai** [comment]
Thanks @rtimms, I don't like the way of making lots of parameter sets (Chen2020) as well, but it will not address this issue by making this model to submodels. Dr. Mayur Bonkile (new postdoc at Imperial) will take it over and forward, for other coupled models.

I will keep this branch locally, and instead of leading to my repository, how about creating a copied branch at the pybamm main repository?

**Comment by rtimms** [comment]
Yes, agreed that making this into submodels won't solve the issue about parameter sets, but having the model more closely integrated would justify adding an extra parameter set now while we figure out a better way to handle parameters.

I was suggesting doing something like [this example](https://github.com/rtimms/dfn-example) where you can define your model and parameters (and specify the version of pybamm it works in). Then you can include links to your paper and people can download and run the model, but it doesn't require maintaining a separate pybamm branch. My worry with just having a branch on the main pybamm repo with this composite model in is that it will get buried away. By having a separate repo it's easier to point people to and easier to google - hopefully people searching your paper can find the repo or you could tweet or similar to point people to it when you advertise the paper.

**Comment by weilongai** [comment]
Thanks @rtimms, I will close this pull request and quote my own repository in the issue #(1433)

**Comment by rtimms** [comment]
Great, thanks! Let us know if you or Mayur want any help in the future integrating this work.
